app/components/filters/select-multiple-filter.ts

```diff
@@ -29,59 +33,61 @@ export default class SelectMultipleFilterComponent extends FilterComponent<Signa
 
   @action
   async inserted() {
-    if (this.args.queryParam?.includes('+')) {
-      const needles = deserializeArray(this.args.queryParam).map(
-        (queryParam) => {
-          if (!queryParam) return [];
-          return this.getQueryParam(queryParam);
-        }
-      );
+    const searchField = this.args.searchField;
 
-      const searchField = this.args.searchField;
+    if (!this.router.currentRoute) {
+      console.error('Current route is not available');
+      return;
+    }
 
-      const haystack = (await this.args.options) as unknown as GroupedOptions[];
-      const results: Option[] = [];
+    const results: Option[] = [];
+    const haystack = (await this.args.options) as unknown as GroupedOptions[];
 
-      const flattenedHaystack: Option[] = [];
-      if (haystack[0]?.['groupName']) {
-        haystack.forEach((group) => {
-          group['options'].forEach((option: Option) => {
-            flattenedHaystack.push(option);
-          });
+    const flattenedHaystack: Option[] = [];
+    if (haystack?.[0]?.['groupName']) {
+      haystack.forEach((group) => {
+        group['options'].forEach((option: Option) => {
+          flattenedHaystack.push(option);
         });
-      }
-
-      needles.forEach((needle) => {
-        const found = flattenedHaystack.find(
-          (value) => get(value, searchField) === needle
-        );
-        if (found) {
-          results.push(found);
-        }
       });
+    }
 
-      this.onSelectedChange(results);
-    } else {
-      const queryParam = this.getQueryParam(this.args.queryParam) as string;
-      if (!queryParam) return [];
-
-      const needles = deserializeArray(queryParam);
-      const searchField = this.args.searchField;
-
-      const haystack = await this.args.options;
-      const results: Option[] = [];
+    const queryParam = this.args.queryParam;
+    const needles = deserializeArray(queryParam)
+      .map((queryParam) => {
+        if (!queryParam) return undefined;
+        const value = this.getQueryParam(queryParam);
+        if (typeof value === 'string') {
+          return value.split('+').map((splitValue) => {
+            return {
+              queryParam: queryParam,
+              value: splitValue,
+            };
+          });
+        } else {
+          return {
+            queryParam: queryParam,
+            value: value,
+          };
+        }
+      })
+      .filter(Boolean)
+      .flat();
 
-      needles.forEach((needle) => {
-        const found = haystack.find(
-          (value) => get(value, searchField) === needle
+    needles.forEach((needle) => {
+      if (needle) {
+        const found = flattenedHaystack.find(
+          (value) =>
+            get(value, searchField) === needle.value &&
+            value['type'] === needle.queryParam
         );
         if (found) {
           results.push(found);
         }
-      });
+      }
+    });
```

Comment:
**Suggestion:** I would like to go further on the simplification by merging the `needles` and `results` loops. 

What do you think about this example?  
```typescript
const results = deserializeArray(this.args.queryParam).flatMap(
  (queryParam) => {
    if (!queryParam) return [];
    const queryParamValue = this.getQueryParam(queryParam);
    const values = queryParamValue ? deserializeArray(queryParamValue) : [];
    return values
      .map((value) => {
        return flattenedHaystack.find(
          (option) =>
            get(option, searchField) === value &&
            option['type'] === queryParam
        );
      })
      .filter(Boolean) as Option[];
  }
);
```